webui: use server presets as parameter placeholders

ServeurpersoCom opened the pull request:
*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

WebUI now displays dynamic parameter placeholders from server/model defaults instead of hardcoded values. Sampling parameters are removed from static defaults - the backend becomes the source of truth while users can still override when needed.

### Changes:
- Remove hardcoded sampling parameters from `SETTING_CONFIG_DEFAULT`
- Add `getParameterPlaceholder()`: model presets -> server defaults -> "none"
- Support both router mode (per-model `/props?model=X`) and single-model mode (`/props`)
- Unify "Custom" parameter badge logic across both modes - badge appears when user value differs from current placeholder

### Behavior:
- When no user override is set, server automatically uses its presets (or defaults if no preset available) -> webui don't send this API field.
- Parameter placeholders are reactive and automatically update when switching models
- Displays current model's preset values from `default_generation_settings.params`
- User overrides are stored globally in localStorage (not per-model)
- Per-model overrides can be implemented as future enhancement

### Demo:
https://github.com/user-attachments/assets/c993aea7-3750-490d-9ded-40276c9a47dd

Close #18129

ServeurpersoCom committed:
webui: use server presets as parameter placeholders

WebUI sampling parameters now default to null and display server
presets (from models.ini or CLI args) as placeholders instead of
hardcoded defaults. This allows the backend to be the source of
truth while users can still override parameters when needed

ServeurpersoCom committed:
webui: normalize floating-point values in parameter placeholders

allozaur commented on tools/server/webui/src/lib/stores/settings.svelte.ts:
```diff
@@ -174,7 +191,14 @@ class SettingsStore {
 				if (propsDefault !== undefined) {
+					if (isUnset) {
+						this.userOverrides.delete(key);
+						continue;
```
```suggestion
						this.userOverrides.delete(key);

						continue;
```

allozaur reviewed (changes requested):
few remarks

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatSettings/ChatSettingsFields.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatSettings/ChatSettingsFields.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/stores/settings.svelte.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/stores/settings.svelte.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/stores/settings.svelte.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/stores/settings.svelte.ts

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
webui: remove null defaults for server-synced parameters

ServeurpersoCom committed:
webui: fix 'Custom' badge logic to work consistently in both router and single-model modes

ServeurpersoCom committed:
webui: refactor all settings to use server defaults only

Remove SETTING_CONFIG_DEFAULT and canSync flag, all settings now
sourced from server /props webui_settings. Admin configures UI
defaults via --webui-config-file. Config starts empty, populated
from server. Placeholder falls back to 'none' when unavailable

ServeurpersoCom commented:
Everything is customizable via --webui-config-file.

The JSON parameters will be:

1) Loaded on all new browsers (first visit)
2) Customizable per browser and stored in localStorage
3) Resettable via "Reset to default" button to restore server defaults

```
{
  "apiKey": "",
  "systemMessage": "",
  "showSystemMessage": true,
  "theme": "system",
  "pasteLongTextToFileLen": 2500,
  "copyTextAttachmentsAsPlainText": false,
  "pdfAsImage": false,
  "showThoughtInProgress": false,
  "showToolCalls": false,
  "disableReasoningFormat": false,
  "alwaysShowSidebarOnDesktop": false,
  "autoShowSidebarOnNewChat": true,
  "keepStatsVisible": false,
  "showMessageStats": true,
  "askForTitleConfirmation": false,
  "disableAutoScroll": false,
  "renderUserContentAsMarkdown": false,
  "autoMicOnEmpty": false,
  "custom": "",
  "pyInterpreterEnabled": false,
  "enableContinueGeneration": false
}
```

ServeurpersoCom commented:
Rebased / static build

ServeurpersoCom committed:
webui: replace legacy 'auto' theme with 'system'

Unify theme naming: remove leftover 'auto' references, use 'system'
consistently (matches UI label and JSON config)

ServeurpersoCom committed:
webui: type-based config fallbacks via parameter metadata

Introspect SYNCABLE_PARAMETERS types for intelligent defaults,
eliminates manual ?? false throughout codebase

ServeurpersoCom committed:
webui: finish the config() -> getConfig() refactor we started

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatForm/ChatForm.svelte:
```diff
@@ -67,7 +66,13 @@
+
+		const placeholder = settingsStore.getParameterPlaceholder('pasteLongTextToFileLen');
+		const placeholderNumber = Number(placeholder);
+		return Number.isNaN(placeholderNumber) ? 0 : placeholderNumber;
```
```suggestion
		const placeholderNumber = Number(placeholder);

		return Number.isNaN(placeholderNumber) ? 0 : placeholderNumber;
```

allozaur reviewed (changes requested):
@ServeurpersoCom overall great stuff, but architecturally i think it'd be good to address the way how we are getting the config into the stores and components.

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatForm/ChatForm.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessageEditForm.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom commented:
There is still work to be done (+ doc update) :
```
(root|~/llama.cpp.pascal/tools/server/webui) git grep "config()"
docs/architecture/high-level-architecture.md:                RE35["config()"]
src/lib/components/app/chat/ChatForm/ChatForm.svelte:   let currentConfig = $derived(config());
src/lib/components/app/chat/ChatForm/ChatFormActions/ChatFormActions.svelte:    let currentConfig = $derived(config());
src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte:   let currentConfig = $derived(config());
src/lib/components/app/chat/ChatMessages/ChatMessageEditForm.svelte:    let currentConfig = $derived(config());
src/lib/components/app/chat/ChatSettings/ChatSettings.svelte:   let localConfig: SettingsConfigType = $state({ ...config() });
src/lib/components/app/chat/ChatSettings/ChatSettings.svelte:           localConfig = { ...config() };
src/lib/components/app/chat/ChatSettings/ChatSettings.svelte:           localConfig = { ...config() };
src/lib/services/chat.ts:                       const agenticConfig = mcpClient ? getAgenticConfig(config()) : undefined;
src/lib/services/mcp-singleton.ts:      const mcpConfig = buildMcpClientConfig(config());
src/lib/services/mcp-singleton.ts:      const mcpConfig = buildMcpClientConfig(config());
src/lib/stores/chat.svelte.ts:          const currentConfig = config();
src/lib/stores/chat.svelte.ts:          const currentConfig = config();
(root|~/llama.cpp.pascal/tools/server/webui) git grep SETTING_CONFIG_DEFAULT
docs/flows/settings-flow.md:        settingsStore->>settingsStore: Merge with SETTING_CONFIG_DEFAULT
docs/flows/settings-flow.md:        settingsStore->>settingsStore: config = SETTING_CONFIG_DEFAULT
docs/flows/settings-flow.md:    settingsStore->>settingsStore: config = SETTING_CONFIG_DEFAULT
```
```
(root|~/llama.cpp.pascal/tools/server/webui) git grep "currentConfig"
src/lib/components/app/chat/ChatForm/ChatForm.svelte:   let currentConfig = $derived(config());
src/lib/components/app/chat/ChatForm/ChatForm.svelte:           const n = Number(currentConfig.pasteLongTextToFileLen);
src/lib/components/app/chat/ChatForm/ChatFormActions/ChatFormActions.svelte:    let currentConfig = $derived(config());
src/lib/components/app/chat/ChatForm/ChatFormActions/ChatFormActions.svelte:            hasAudioModality && !hasText && !hasAudioAttachments && currentConfig.autoMicOnEmpty
src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte:   let currentConfig = $derived(config());
src/lib/components/app/chat/ChatMessages/ChatMessageAssistant.svelte:                           {:else if isLoading() && currentConfig.showMessageStats}
src/lib/components/app/chat/ChatMessages/ChatMessageEditForm.svelte:    let currentConfig = $derived(config());
src/lib/components/app/chat/ChatMessages/ChatMessageEditForm.svelte:            const n = Number(currentConfig.pasteLongTextToFileLen);
src/lib/stores/chat.svelte.ts:          const currentConfig = config();
src/lib/stores/chat.svelte.ts:                  temperature: currentConfig.temperature ?? 0.8,
src/lib/stores/chat.svelte.ts:                  topP: currentConfig.top_p ?? 0.95,
src/lib/stores/chat.svelte.ts:          const currentConfig = config();
src/lib/stores/chat.svelte.ts:          if (currentConfig.disableReasoningFormat) apiOptions.disableReasoningFormat = true;
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.temperature))
src/lib/stores/chat.svelte.ts:                  apiOptions.temperature = Number(currentConfig.temperature);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.dynatemp_range))
src/lib/stores/chat.svelte.ts:                  apiOptions.dynatemp_range = Number(currentConfig.dynatemp_range);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.dynatemp_exponent))
src/lib/stores/chat.svelte.ts:                  apiOptions.dynatemp_exponent = Number(currentConfig.dynatemp_exponent);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.top_k)) apiOptions.top_k = Number(currentConfig.top_k);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.top_p)) apiOptions.top_p = Number(currentConfig.top_p);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.min_p)) apiOptions.min_p = Number(currentConfig.min_p);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.xtc_probability))
src/lib/stores/chat.svelte.ts:                  apiOptions.xtc_probability = Number(currentConfig.xtc_probability);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.xtc_threshold))
src/lib/stores/chat.svelte.ts:                  apiOptions.xtc_threshold = Number(currentConfig.xtc_threshold);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.typ_p)) apiOptions.typ_p = Number(currentConfig.typ_p);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.repeat_last_n))
src/lib/stores/chat.svelte.ts:                  apiOptions.repeat_last_n = Number(currentConfig.repeat_last_n);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.repeat_penalty))
src/lib/stores/chat.svelte.ts:                  apiOptions.repeat_penalty = Number(currentConfig.repeat_penalty);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.presence_penalty))
src/lib/stores/chat.svelte.ts:                  apiOptions.presence_penalty = Number(currentConfig.presence_penalty);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.frequency_penalty))
src/lib/stores/chat.svelte.ts:                  apiOptions.frequency_penalty = Number(currentConfig.frequency_penalty);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.dry_multiplier))
src/lib/stores/chat.svelte.ts:                  apiOptions.dry_multiplier = Number(currentConfig.dry_multiplier);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.dry_base)) apiOptions.dry_base = Number(currentConfig.dry_base);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.dry_allowed_length))
src/lib/stores/chat.svelte.ts:                  apiOptions.dry_allowed_length = Number(currentConfig.dry_allowed_length);
src/lib/stores/chat.svelte.ts:          if (hasValue(currentConfig.dry_penalty_last_n))
src/lib/stores/chat.svelte.ts:                  apiOptions.dry_penalty_last_n = Number(currentConfig.dry_penalty_last_n);
src/lib/stores/chat.svelte.ts:          if (currentConfig.samplers) apiOptions.samplers = currentConfig.samplers;
src/lib/stores/chat.svelte.ts:          if (currentConfig.custom) apiOptions.custom = currentConfig.custom;
src/lib/utils/process-uploaded-files.ts:                                const currentConfig = settingsStore.config;
src/lib/utils/process-uploaded-files.ts:                                if (hasVisionSupport && !currentConfig.pdfAsImage) {
```

ServeurpersoCom committed:
webui: move normalizeComparisonValue to utils/precision

Address review feedback from allozaur

ServeurpersoCom committed:
webui: remove unnecessary Boolean() wrappers

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatScreen/ChatScreen.svelte:
```diff
@@ -41,7 +41,7 @@
 	let { showCenteredEmpty = false } = $props();
 
-	let disableAutoScroll = $derived(Boolean(config().disableAutoScroll));
+	let disableAutoScroll = $derived(Boolean(settingsStore.getConfig('disableAutoScroll')));
```
Yep, let's try this. It's just that I want us to have this as clean as possible 😊

ServeurpersoCom replied:
Yes, me too!!!

ServeurpersoCom replied:
Everything works fine

ServeurpersoCom committed:
webui: migrate simple components from config() to getConfig()

ServeurpersoCom committed:
npm run format

ServeurpersoCom committed:
webui: refactor getApiOptions and remove dead code

- Clean up unused config imports
- Refactor getApiOptions: replace 40-line if-chain with loop
- Remove dead ApiProcessingState fields (temperature, topP, speculative)

ServeurpersoCom committed:
webui: migrate ChatSettings from config() to settingsStore.config

ServeurpersoCom committed:
webui: remove unused config() export

ServeurpersoCom committed:
webui: migrate process-uploaded-files to getConfig()

ServeurpersoCom committed:
chore: update webui build output

ServeurpersoCom commented on tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessages.svelte:
```diff
@@ -43,7 +42,7 @@
 
 		// Filter out system messages if showSystemMessage is false
-		const filteredMessages = currentConfig.showSystemMessage
+		const filteredMessages = settingsStore.getConfig('showSystemMessage')
```
Done
